### PR TITLE
[Auth] Implement user registration endpoint

### LINKED ISSUE #3: [Auth] Implement user registration endpoint
## Description
Create the POST /api/auth/register endpoint for new user registration.

## Acceptance Criteria
- [ ] Create RegisterDTO with validation (email, username, password, displayName)
- [ ] Implement password hashing with bcrypt
- [ ] Check for duplicate email/username
- [ ] Return user data (without password) on success
- [ ] Add proper error handling

## Category
`auth`

## Depends on
- #2 User Entity

### PR DESCRIPTION
**Pull request by math97**
## Summary
- Implements POST `/api/auth/register` endpoint for new user registration
- Adds RegisterDTO with validation (email, username, password, displayName)
- Implements password hashing with bcrypt
- Adds duplicate email/username checking with proper error handling
- Creates PrismaUserRepository and PrismaService for database operations

## Test plan
- [x] Test successful user registration with valid data
- [ ] Test validation errors for invalid email/username/password formats
- [ ] Test conflict errors for duplicate email
- [ ] Test conflict errors for duplicate username
- [ ] Verify password is hashed and not returned in response

Closes #3

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -1,9 +1,17 @@
...
+    new ValidationPipe({
+      whitelist: true,
+      forbidNonWhitelisted: true,
+      transform: true,
+    }),
+  );
   await app.listen(process.env.PORT ?? 3000);
 }
 
-bootstrap();
```
Comment (math97) [inline src/main.ts:9]: does not need to add this void here

File: src/infra/modules/auth/dtos/register.dto.ts
Comment (math97) [inline src/infra/modules/auth/dtos/register.dto.ts:1]: we shoud use zod and implement a custom pipeline for validation on http level

**Comment by math97** [comment]
## PR Review

### Summary
This PR implements user registration functionality for the authentication system. It adds a `POST /api/auth/register` endpoint with proper validation, password hashing using bcrypt, duplicate checking for email/username, and establishes the foundational database layer with Prisma.

### Findings

#### Issues

1. **Missing `url` in Prisma datasource** (`prisma/schema.prisma`)
   - The `url` property was removed from the datasource block, which may cause issues with Prisma CLI commands (migrations, generate). While the `prisma.config.ts` sets the URL programmatically, the schema should still have `url = env("DATABASE_URL")` for CLI compatibility.

2. **PrismaService directly accesses `process.env`** (`src/infra/database/prisma.service.ts:11`)
   - The `DATABASE_URL` is accessed directly via `process.env` instead of using NestJS's `ConfigService`. This bypasses the `EnvModule` validation and could fail silently if the variable is undefined.

3. **No unit tests included**
   - The PR description includes a test plan but no actual tests are implemented. Consider adding unit tests for the `AuthService` (especially the duplicate checking logic) and integration tests for the endpoint.

#### Suggestions

1. **Consider normalizing email before storage**
   - Emails could be stored in lowercase to prevent duplicate accounts with different casing (e.g., `User@Email.com` vs `user@email.com`).

2. **Add password strength validation**
   - The current validation only checks for minimum 8 characters. Consider adding requirements for complexity (uppercase, lowercase, numbers, special characters) to improve security.

3. **Consider using a constant for bcrypt salt rounds**
   - The `10` in `bcrypt.hash(dto.password, 10)` could be extracted to a constant or configuration value for easier maintenance.

4. **The `@types/bcryptjs` package shows a deprecation warning**
   - The lock file shows this package is deprecated since `bcryptjs` now includes its own types. You can remove `@types/bcryptjs` from devDependencies.

5. **Consider adding rate limiting**
   - Registration endpoints are common targets for abuse. Consider implementing rate limiting to prevent mass account creation.

#### Positive Notes

- Clean separation of concerns with domain entities, repository interfaces, and infrastructure implementations following clean architecture principles
- Good use of dependency injection with the `USER_REPOSITORY` token pattern, making the service testable
- Proper use of NestJS's `ValidationPipe` with `whitelist` and `forbidNonWhitelisted` options for input sanitization
- Password is correctly excluded from the response using destructuring
- Database schema uses proper naming conventions with `@map` for snake_case columns while keeping camelCase in the code
- Migration file is properly versioned and creates appropriate unique constraints

### Conclusion
This is a solid foundation for the authentication system. The main concerns are the missing `url` in the Prisma schema and the direct `process.env` access in `PrismaService`. These should be addressed before merging. Adding tests would also strengthen confidence in the implementation.

---
*Reviewed by Claude Code*

**Comment by math97** [comment]
## PR Review

### Summary
This PR implements the user registration endpoint (`POST /api/auth/register`) for the authentication system. It includes:
- User entity and repository abstraction layer
- Prisma database integration with PostgreSQL
- Password hashing with bcrypt
- Input validation using Zod schemas
- Duplicate email/username checking

### Findings

#### Issues
1. **Missing validation pipe in controller (in diff)**: The diff shows the controller without `@UsePipes(new ZodValidationPipe(registerSchema))`, but the current file has it. This discrepancy suggests the PR may have been updated after the diff was generated, or the validation is being added elsewhere. Verify that validation is actually applied to the `/register` endpoint.

2. **bcrypt cost factor**: The bcrypt cost factor of 10 is acceptable, but consider making it configurable via environment variable for future flexibility.

3. **Missing rate limiting**: The registration endpoint should have rate limiting to prevent abuse (mass account creation). Consider adding a throttler guard.

4. **Test coverage incomplete**: The PR description shows several test cases as unchecked:
   - [ ] Test validation errors for invalid email/username/password formats
   - [ ] Test conflict errors for duplicate email
   - [ ] Test conflict errors for duplicate username
   - [ ] Verify password is hashed and not returned in response

#### Suggestions
1. **Add email normalization**: Consider lowercasing emails before storing to prevent duplicate accounts with different casing (e.g., `User@Email.com` vs `user@email.com`).

2. **Username normalization**: Similar to email, consider whether usernames should be case-insensitive for lookups.

3. **Password strength validation**: The current minimum is 8 characters. Consider adding requirements for complexity (uppercase, lowercase, numbers, special characters) or using a library like `zxcvbn` for password strength estimation.

4. **Response type safety**: Consider creating a proper DTO for the registration response to ensure `passwordHash` is never accidentally included.

5. **Database module extraction**: The `PrismaService` and `PrismaUserRepository` could be extracted into a dedicated `DatabaseModule` for better separation of concerns and reusability across other modules.

#### Positive Notes
- Clean architecture with proper separation between domain entities, repositories, and infrastructure
- Good use of dependency injection with the `USER_REPOSITORY` token pattern
- Proper use of Zod for runtime validation
- Password hash is correctly excluded from the response
- Proper lifecycle hooks (`onModuleInit`, `onModuleDestroy`) for database connection management
- The migration SQL is well-structured with appropriate indexes on `email` and `username`

### Conclusion
The implementation follows good practices and provides a solid foundation for the authentication system. The main concerns are around missing test coverage and the lack of rate limiting on the registration endpoint. Consider addressing the security suggestions (rate limiting, email normalization) before merging, and ensure all test cases in the PR description are implemented.

---
*Reviewed by Claude Code*
